### PR TITLE
Opsrc-355/Add ability to add agreements to each form in Sylius

### PR DESCRIPTION
**Pull request by liszkapawel**
This PR introduces the function of adding agreements to each form in Sylius (extendable forms are user-defined in a configuration file)

### CODE REVIEW

**Comment by jakubtobiasz** [comment]
You've added a totally new feature and no new tests (both unit and behats). As it's open source project we need to provide quality assurance.

File: src/EventSubscriber/AgreementSubscriber.php [hunk 1]
```diff
@@ -10,146 +10,56 @@
...
-                null !== $agreementHistory->getId()
-            ) {
-                $agreementHistory = clone $agreementHistory;
-            }
-
-            $agreementHistory->setState($agreementHistoryState);
-            $this->agreementHistoryRepository->add($agreementHistory);
+    public function processAgreementsFromAnywhere(AgreementCheckedEvent $agreementCheckedEvent): void
+    {
+        if (null === $agreementCheckedEvent->getEvent()->getData()->getUser()->getId()) {
```
Comment (senghe) [inline src/EventSubscriber/AgreementSubscriber.php:40]: Could you put the getter's train into the `AgreementCheckedEvent` class?

File: spec/Resolver/CompositeAgreementApprovalResolverSpec.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            $agreementHistoryResolver
+        );
+    }
+
+    function it_is_initializable(): void
+    {
+        $this->shouldHaveType(CompositeAgreementApprovalResolver::class);
+    }
+
+    function it_resolves_agreement_correctly(
```
Comment (senghe) [inline spec/Resolver/CompositeAgreementApprovalResolverSpec.php:34]: One scenario is missing :)

File: spec/Resolver/CompositeAgreementHistoryResolverSpec.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+use BitBag\SyliusAgreementPlugin\Repository\AgreementHistoryRepositoryInterface;
+use BitBag\SyliusAgreementPlugin\Resolver\CompositeAgreementHistoryResolver;
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Order\Context\CartContextInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Symfony\Component\Security\Core\Security;
+
+final class CompositeAgreementHistoryResolverSpec extends ObjectBehavior
```
Comment (senghe) [inline spec/Resolver/CompositeAgreementHistoryResolverSpec.php:24]: One scenario is missing :)

File: src/Resolver/CompositeAgreementResolver.php [hunk 4]
```diff
@@ -10,31 +10,19 @@
 
 namespace BitBag\SyliusAgreementPlugin\Resolver;
 
+use BitBag\SyliusAgreementPlugin\Repository\AgreementRepositoryInterface;
+
 final class CompositeAgreementResolver implements AgreementResolverInterface
```
Comment (senghe) [inline src/Resolver/CompositeAgreementResolver.php:15]: I think we can remove this class

File: spec/Resolver/CompositeAgreementResolverSpec.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+declare(strict_types=1);
+
+namespace spec\BitBag\SyliusAgreementPlugin\Resolver;
+
+use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementInterface;
+use BitBag\SyliusAgreementPlugin\Repository\AgreementRepositoryInterface;
+use BitBag\SyliusAgreementPlugin\Resolver\CompositeAgreementResolver;
+use PhpSpec\ObjectBehavior;
+
+final class CompositeAgreementResolverSpec extends ObjectBehavior
```
Comment (senghe) [inline spec/Resolver/CompositeAgreementResolverSpec.php:18]: This class also can be removed :)

File: src/DependencyInjection/AgreementFormExtension.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+{
+    public function process(ContainerBuilder $container): void
+    {
+        /** @var array $contexts */
+        $contexts = $container->getParameter('sylius_agreement_plugin.contexts');
+
+        foreach ($contexts as $types) {
+            foreach ($types as $type) {
+                $service = new Definition(AgreementsTypeExtension::class);
+                $id = sprintf('bitbag_sylius_agreement_plugin.form.extension.agreements_type_extension_%s', str_replace(['/', '\\'], '_', $type));
```
Comment (senghe) [inline src/DependencyInjection/AgreementFormExtension.php:29]: Can we put the string to const?

File: src/Entity/Agreement/AgreementHistory.php [hunk 7]
```diff
@@ -33,7 +33,7 @@ class AgreementHistory implements AgreementHistoryInterface
     protected $state = AgreementHistoryStates::STATE_ASSIGNED;
 
     /** @var string */
-    protected $context = AgreementContexts::CONTEXT_UNKNOWN;
+    protected $context = 'unknown';
```
Comment (senghe) [inline src/Entity/Agreement/AgreementHistory.php:36]: Could you put the const into current class?

File: src/Form/Type/Agreement/Admin/AgreementType.php [hunk 8]
```diff
@@ -60,8 +60,8 @@ protected function prepareContextsData(): array
     {
         $contexts = [];
 
-        foreach ($this->contexts as $context) {
-            $contexts[\sprintf('bitbag_sylius_agreement_plugin.ui.agreement.contexts.%s', $context)] = $context;
+        foreach ($this->contexts as $key => $val) {
+            $contexts[\sprintf('bitbag_sylius_agreement_plugin.ui.agreement.contexts.%s', $key)] = $key;
```
Comment (senghe) [inline src/Form/Type/Agreement/Admin/AgreementType.php:64]: Could you put string into const?

File: src/Handler/AgreementHandler.php [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    private function getSubmittedAgreement(Collection $submittedAgreements, AgreementInterface $resolvedAgreement): AgreementInterface
+    {
+        return $submittedAgreements->filter(
+            static function (AgreementInterface $agreement) use ($resolvedAgreement): bool {
+                return $agreement->getId() === $resolvedAgreement->getId();
+            }
+        )->first();
+    }
+
+    private function setAgreementHistoryProperties(
```
Comment (senghe) [inline src/Handler/AgreementHandler.php:104]: Can we put this into separate class?

File: src/Handler/AgreementHandler.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+            AgreementHistoryStates::STATE_SHOWN !== $resolvedAgreementHistoryState
+            && null !== $agreementHistory->getId()
+        ) {
+            $agreementHistoryState = AgreementHistoryStates::STATE_WITHDRAWN;
+        }
+
+        return $agreementHistoryState;
+    }
+
+    private function getSubmittedAgreement(Collection $submittedAgreements, AgreementInterface $resolvedAgreement): AgreementInterface
```
Comment (senghe) [inline src/Handler/AgreementHandler.php:95]: We can move it into new filter class I think

File: src/Handler/AgreementHandler.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            ) {
+                $agreementHistory = clone $agreementHistory;
+            }
+
+            $agreementHistory->setState($agreementHistoryState);
+            $this->agreementHistoryRepository->add($agreementHistory);
+        }
+    }
+
+    private function determineState(
```
Comment (senghe) [inline src/Handler/AgreementHandler.php:76]: It can be a separate class

File: src/Resolver/CompositeAgreementApprovalResolver.php [hunk 12]
```diff
@@ -10,32 +10,27 @@
 
 namespace BitBag\SyliusAgreementPlugin\Resolver;
 
+use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementHistoryInterface;
+use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementHistoryStates;
 use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementInterface;
-use BitBag\SyliusAgreementPlugin\Exception\AgreementNotSupportedException;
 
 final class CompositeAgreementApprovalResolver implements AgreementApprovalResolverInterface
```
Comment (senghe) [inline src/Resolver/CompositeAgreementApprovalResolver.php:17]: It's checker than resolver :P

File: src/Resolver/CompositeAgreementHistoryResolver.php [hunk 13]
```diff
@@ -12,45 +12,54 @@
...
-        if (!$agreementHistory instanceof AgreementHistoryInterface) {
+        /** @var OrderInterface|null $order */
+        $order = $this->cartContext->getCart();
+
+        /** @var ShopUserInterface|null $shopUser */
+        $shopUser = $this->security->getUser();
+
+        if (null !== $shopUser) {
+            /** @var AgreementHistoryInterface|null $agreementHistory */
+            $agreementHistory = $this->agreementHistoryRepository->findOneForShopUser($agreement, $shopUser);
```
Comment (senghe) [inline src/Resolver/CompositeAgreementHistoryResolver.php:54]: We have to check all returned values wether they're null and if so, we should return a newly created history - I think

File: tests/Application/src/Resources/config/doctrine/Order.Order.orm.xml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
+                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                  xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping
+                                      http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd">
+    <mapped-superclass name="Tests\BitBag\SyliusAgreementPlugin\Entity\Order\Order"
```
Comment (senghe) [inline tests/Application/src/Resources/config/doctrine/Order.Order.orm.xml:7]: I think better here would be entity than mapped-superclass

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/Resolver/CompositeAgreementResolver.php (hunk 4 above)
Comment (liszkapawel) [reply src/Resolver/CompositeAgreementResolver.php]: Done

File: src/DependencyInjection/AgreementFormExtension.php (hunk 6 above)
Comment (liszkapawel) [reply src/DependencyInjection/AgreementFormExtension.php:29]: Done

File: src/Form/Type/Agreement/Admin/AgreementType.php (hunk 8 above)
Comment (liszkapawel) [reply src/Form/Type/Agreement/Admin/AgreementType.php:64]: Done

File: tests/Application/src/Resources/config/doctrine/Order.Order.orm.xml (hunk 14 above)
Comment (liszkapawel) [reply tests/Application/src/Resources/config/doctrine/Order.Order.orm.xml]: Done

File: src/Resolver/CompositeAgreementApprovalResolver.php (hunk 12 above)
Comment (liszkapawel) [reply src/Resolver/CompositeAgreementApprovalResolver.php]: Done

File: spec/Resolver/CompositeAgreementApprovalResolverSpec.php (hunk 2 above)
Comment (liszkapawel) [reply spec/Resolver/CompositeAgreementApprovalResolverSpec.php]: Done

File: spec/Resolver/CompositeAgreementHistoryResolverSpec.php (hunk 3 above)
Comment (liszkapawel) [reply spec/Resolver/CompositeAgreementHistoryResolverSpec.php:24]: Done

Comment (liszkapawel) [reply spec/Resolver/CompositeAgreementHistoryResolverSpec.php:24]: Function is changed, no more scenario is needed

File: src/Resolver/CompositeAgreementHistoryResolver.php (hunk 13 above)
Comment (liszkapawel) [reply src/Resolver/CompositeAgreementHistoryResolver.php:54]: Changed

File: src/Handler/AgreementHandler.php (hunk 10 above)
Comment (liszkapawel) [reply src/Handler/AgreementHandler.php:95]: I put the rest of the functions into other classes, I think we can leave it here, we won't test it anyway

File: src/Handler/AgreementHandler.php (hunk 11 above)
Comment (liszkapawel) [reply src/Handler/AgreementHandler.php:76]: Done

File: src/Handler/AgreementHandler.php (hunk 9 above)
Comment (liszkapawel) [reply src/Handler/AgreementHandler.php:104]: Done, I create DataModifier class but I'm not sure that's the right name

File: spec/EventSubscriber/AgreementSubscriberSpec.php [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+use BitBag\SyliusAgreementPlugin\Event\AgreementCheckedEvent;
+use BitBag\SyliusAgreementPlugin\EventSubscriber\AgreementSubscriber;
+use BitBag\SyliusAgreementPlugin\Handler\AgreementHandler;
+use Doctrine\Common\Collections\Collection;
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Symfony\Component\Form\FormEvent;
+use Tests\BitBag\SyliusAgreementPlugin\Entity\Order\OrderInterface;
+
+final class AgreementSubscriberSpec extends ObjectBehavior
```
Comment (senghe) [inline spec/EventSubscriber/AgreementSubscriberSpec.php:22]: Two scenarios are missing

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/EventSubscriber/AgreementSubscriberSpec.php (hunk 15 above)
Comment (liszkapawel) [reply spec/EventSubscriber/AgreementSubscriberSpec.php:22]: Done

File: src/Handler/AgreementHandler.php [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+                $agreementHistoryState !== $resolvedAgreementHistoryState &&
+                null !== $agreementHistory->getId()
+            ) {
+                $agreementHistory = clone $agreementHistory;
+            }
+
+            $agreementHistory->setState($agreementHistoryState);
+            $this->agreementHistoryRepository->add($agreementHistory);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (liszkapawel) [reply src/Handler/AgreementHandler.php:76]: Done

File: spec/Resolver/AgreementStateResolverSpec.php [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+namespace spec\BitBag\SyliusAgreementPlugin\Resolver;
+
+use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementHistoryInterface;
+use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementHistoryStates;
+use BitBag\SyliusAgreementPlugin\Entity\Agreement\AgreementInterface;
+use BitBag\SyliusAgreementPlugin\Resolver\AgreementStateResolver;
+use PhpSpec\ObjectBehavior;
+
+
+final class AgreementStateResolverSpec extends ObjectBehavior
```
[the remark this answers is not in the document]
Comment (liszkapawel) [reply spec/Resolver/AgreementStateResolverSpec.php:20]: Done

File: spec/EventSubscriber/AgreementSubscriberSpec.php [hunk 18]
```diff
@@ -68,6 +69,61 @@ function it_process_successfully(
             ->shouldBeCalled();
 
         $this->processAgreementsFromAnywhere($agreementCheckedEvent);
+    }
+
+    function it_process_successfully_when_order_id_is_null(
```
Comment (senghe) [inline spec/EventSubscriber/AgreementSubscriberSpec.php:74]:
```suggestion
    function it_processes_successfully_when_order_id_is_null(
```

File: spec/EventSubscriber/AgreementSubscriberSpec.php [hunk 19]
```diff
@@ -68,6 +69,61 @@ function it_process_successfully(
...
+            $userAgreements,
+            'checkout_form',
+            null,
+            $shopUser)
+            ->shouldBeCalled();
 
+        $this->processAgreementsFromAnywhere($agreementCheckedEvent);
+    }
+
+    function it_process_successfully_when_data_is_not_instance_of_order(
```
Comment (senghe) [inline spec/EventSubscriber/AgreementSubscriberSpec.php:101]:
```suggestion
    function it_processes_successfully_when_data_is_not_instance_of_order(
```

File: spec/Handler/AgreementHandlerSpec.php [hunk 20]
```diff
@@ -80,4 +81,86 @@ function it_handle_agreements_correctly(
...
+
+    function it_should_throw_exception_when_no_agreements_returned(
+        Collection $collection,
+        OrderInterface $order,
+        ShopUserInterface $shopUser,
+        AgreementRepositoryInterface $agreementRepository
+    ): void {
+        $agreementRepository->findAgreementsByContext('checkout_form')->willReturn([]);
+
+        $this->shouldThrow(NotFoundHttpException::class)
```
Comment (senghe) [inline spec/Handler/AgreementHandlerSpec.php:93]: Why throwing HTTP exception inside bussines logic? I think we should throw dedicated exception for bussines logic and then catch it in controller and then throw not found

File: spec/Handler/AgreementHandlerSpec.php [hunk 21]
```diff
@@ -80,4 +81,86 @@ function it_handle_agreements_correctly(
...
+        )->willReturn(AgreementHistoryStates::STATE_ACCEPTED);
+        $agreementHistory->getState()->willReturn(AgreementHistoryStates::STATE_ASSIGNED);
+        $agreementHistory->getId()->willReturn(null);
+
+        $agreementHistory->setState('accepted')->shouldBeCalled();
+
+        $this->handleAgreements($collection, 'checkout_form', $order, $shopUser);
+    }
+
+    function it_handle_agreements_correctly_when_state_is_other(
```
Comment (senghe) [inline spec/Handler/AgreementHandlerSpec.php:132]:
```suggestion
    function it_handles_agreements_correctly_on_different_state(
```

File: src/Handler/AgreementHandler.php [hunk 22]
```diff
@@ -49,6 +50,10 @@ public function handleAgreements(
     ): void {
         $resolvedAgreements = $this->agreementRepository->findAgreementsByContext($context);
 
+        if (empty($resolvedAgreements)) {
```
Comment (senghe) [inline src/Handler/AgreementHandler.php:53]: We shouldn't use `empty()` function

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/EventSubscriber/AgreementSubscriberSpec.php (hunk 18 above)
Comment (liszkapawel) [reply spec/EventSubscriber/AgreementSubscriberSpec.php:74]: Done

File: spec/EventSubscriber/AgreementSubscriberSpec.php (hunk 19 above)
Comment (liszkapawel) [reply spec/EventSubscriber/AgreementSubscriberSpec.php:101]: Done

File: spec/Handler/AgreementHandlerSpec.php (hunk 21 above)
Comment (liszkapawel) [reply spec/Handler/AgreementHandlerSpec.php:132]: Done

File: src/Handler/AgreementHandler.php (hunk 22 above)
Comment (liszkapawel) [reply src/Handler/AgreementHandler.php:53]: Changed to count.

File: spec/Handler/AgreementHandlerSpec.php (hunk 20 above)
Comment (senghe) [reply spec/Handler/AgreementHandlerSpec.php:93]: We decided to not to do this as we have no custom controller for the form handling.

**Review by senghe — APPROVED** [review]

**Pull request merged**